Create rule S6912: Use batch Processing in JDBC (#3612)
SonarSource/rspec: rules/S6912/java/rule.adoc

== Why is this an issue?

Frequent communication with the database can introduce overhead and decrease performance, especially in scenarios involving multiple SQL statements.

> By batching SQL queries, the number of interactions with the database is reduced, leading to improved efficiency and faster execution times.

Reviewer: Normally I prefer not to start a sentence with "By". What about the following?
```suggestion
Batching SQL queries is a great way to reduce the number of interactions with the database hence improving the efficiency and the execution times.
```

> This can be achieved using the methods provided by the `java.sql.Statement` and `java.sql.PreparedStatement` interfaces for batch processing.

Reviewer: I would add an empty line and rephrase a bit as follows:

```suggestion

The `java.sql.Statement` and `java.sql.PreparedStatement` interfaces can be used for batch processing.
```

What do you think? Is this appropriate for the Why is this an issue section?

=== Code examples

==== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----
public void execute(Connection connection) {
  try {
    Statement statement = connection.createStatement();

    statement.execute("INSERT INTO myTable (column1, column2) VALUES (123, 'value1')"); // Noncompliant
    statement.execute("INSERT INTO myTable (column1, column2) VALUES (456, 'value2')"); // Noncompliant
    statement.execute("INSERT INTO myTable (column1, column2) VALUES (789, 'value3')"); // Noncompliant

    statement.close();
    connection.close();
  } catch (SQLException e) {
    e.printStackTrace();
  }
}
----

==== Compliant solution

[source,java,diff-id=1,diff-type=compliant]
----
public void execute(Connection connection) {
  try {
    Statement statement = connection.createStatement();

    statement.addBatch("INSERT INTO myTable (column1, column2) VALUES (123, 'value1')"); // Compliant
    statement.addBatch("INSERT INTO myTable (column1, column2) VALUES (456, 'value2')"); // Compliant
    statement.addBatch("INSERT INTO myTable (column1, column2) VALUES (789, 'value3')"); // Compliant

    statement.executeBatch();

>     statement.executeBatch();

Reviewer: This seems to be redundant. 

```suggestion
```

I guess that the rule should cover cases in which single statements are executed in a loop. Should we have that as an example case as well? Or mention it somewhere?

Reviewer 2: We have a different rule for that: https://sonarsource.atlassian.net/browse/SONARJAVA-4796
